ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,407 @@
+            // At this point the process is created and we have a pid so
+            // we cannot just return an error.
+            // We could try to wait the process but it should fail, let
+            // returns success and let caller read process status result.
+            return 0;
+        }
+        res->msg.msg_buf[sizeof(res->msg.msg_buf) - 1] = 0;
+        if (readed > 0) {
+            // Wait the process otherwise we'll have a zombie
+            reap_pid(res->pid);
+
+            res->err_msg = res->msg.msg_buf;
+            return res->msg.err;
+        }
+    }
+    return 0;
+}
+
+CAMLprim value
+caml_safe_exec_with_helper(value args, value environment)
+{
+    CAMLparam2(args, environment);
+
+    // Copy parameters to C
+    char **c_args = copy_string_list(args);
+    char **c_envs = copy_string_list(environment);
+    if (!c_envs || !c_args) {
+        free(c_envs);
+        free(c_args);
+        caml_raise_out_of_memory();
+    }
+
+    // potentially slow section, release Ocaml engine
+    caml_enter_blocking_section();
+
+    safe_exec_result res;
+    int err = safe_exec_with_helper(&res, c_args, c_envs);
+
+    free(c_envs);
+    free(c_args);
+
+    caml_leave_blocking_section();
+
+    // error, notify with an exception
+    if (err != 0)
+        unix_error(err, res.err_msg, Nothing);
+
+    CAMLreturn(Val_int(res.pid));
+}
+
+CAMLprim value
+caml_pidwaiter_dontwait(value pid_val)
+{
+    CAMLparam1(pid_val);
+    pid_t pid = Int_val(pid_val);
+
+    // reap the pid to avoid zombies
+    pthread_t th;
+    if (create_thread_minstack(&th, thread_proc_reap, (void *) (intptr_t) pid) == 0)
+        pthread_detach(th);
+
+    CAMLreturn(Val_unit);
+}
+
+typedef struct {
+    pid_t pid;
+    bool timed_out;
+    bool stop;
+    struct timespec deadline;
+    pthread_mutex_t mtx;
+    pthread_cond_t cond;
+} timeout_kill;
+
+static void *
+thread_proc_timeout_kill(void *arg)
+{
+    timeout_kill *tm = (timeout_kill *) arg;
+
+    pthread_mutex_lock(&tm->mtx);
+    int res  = tm->stop ? 0:
+        pthread_cond_timedwait(&tm->cond, &tm->mtx, &tm->deadline);
+    pthread_mutex_unlock(&tm->mtx);
+
+    if (res == ETIMEDOUT) {
+        kill(tm->pid, SIGKILL);
+        tm->timed_out = true;
+    }
+    return NULL;
+}
+
+static int
+create_thread_minstack(pthread_t *th, void *(*proc)(void *), void *arg)
+{
+    int res;
+
+    // disable any possible signal handler so we can safely use a small stack
+    // for the thread
+    sigset_t sigset, old_sigset;
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    pthread_attr_t th_attr;
+    res = pthread_attr_init(&th_attr);
+    if (res)
+        return res;
+    pthread_attr_setstacksize(&th_attr, PTHREAD_STACK_MIN);
+
+    res = pthread_create(th, &th_attr, proc, arg);
+
+    pthread_attr_destroy(&th_attr);
+    pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
+
+    return res;
+}
+
+/*
+ * Wait a process with a given timeout.
```

Comment:
It could, but it was done in C for resource constraints. Launching a thread in higher level languages is not that light (note that there are additional code even to reduce C resource usage).

Reply:
The reason about the thread is not that easy to grasp.

On newer Linux kernel (similarly to Windows) you could use `pidfd` and a simple `poll` to wait for a process easily (on Windows you use the handle to the process). But that would require a newer kernel with newer Glibc (and Linux, but that should not be an issue). Waiting a process without `pidfd` and threads would require some global setting (like a signal handler) which, done in a library, is not nice (library should not change global settings) and racy.

Start to think this would be nice as a comment in the code...

Reply:
Where do these threads run? We wouldn't want to spawn additional C threads inside XAPI, if anything goes wrong there it could crash XAPI. Can we move the monitoring threads to `forkexecd` and query its status via the socket? (although if the spawned process is not a child of forkexec, monitoring its status might be more difficult).

In fact in the future we probably want to move most long running processes to systemd transient units (we just need a newer systemd that supports file descriptor passing), and we can leave the monitoring (and logging) to systemd.

We could also have a single dedicated OCaml thread in XAPI, that is responsible for waiting for all children and notifying the appropriate condition variables. In fact would calling `Unix.wait` in a loop on a dedicated thread achieve that? It wouldn't increase resource usage by much (a single thread should have a fixed cost), in fact with lots of processes it might even decrease the resources used compared to launching a C thread for each one.

This should then further reduce the size of the C code.

Reply:
sorry, your comment does not make sense, the intention is to remove `forkexed`, it always has been, not to use it in different ways.

Reply:
Having a dedicated thread in XAPI doesn't require retaining forkexecd, these were 2 separate suggestions:
* keep the monitoring and vforking in forkexecd if that simplifies the code
* use a dedicated thread in XAPI to avoid the need for extra C code

We can do the latter without doing the former.

Reply:
That explanation makes sense. Your main target is reducing C code. Specifically, this part waiting for a process, potentially with a timeout. However, my Ocaml skills are not that great, so I would probably end up with pretty bad Ocaml code. Can you propose something?

Reply:
I can look in more detail on Friday, in principle something like this (but untested, and probably needs a bit more error handling):
```ocaml
let wait_pid_with_timeout ~timeout pid =
  let need_to_kill = Atomic.make true in
  let kill_pid () =
     if Atomic.get need_to_kill then
     try Unix.kill Sys.SIGKILL pid
    with (Unix.Unix_error(Unix.ESRCH, _, _) -> ()
  in
  let some_unique_id = .... in
  let finally () =  Xapi_periodic_scheduler.remove_from_queue some_unique_id in
  Xapi_periodic_scheduler.add_to_queue some_unique_id Xapi_periodic_scheduler.OneShot timeout kill_pid;
  Fun.protect ~finally (fun () ->
    let pid, status = Unix.waitpid [] pid in
    Atomic.set need_to_kill false;
    status
  )
```

This has the race condition between a process exiting and the timeout kicking in just at the same time (although it is unlikely that we would've also went through all pids, rolled back and got something else meanwhile in that spot).
I see you improved that in your C code, so we could perhaps add a small binding to Unixext.waitid to waitid, so that we can use `waitid` instead of `waitpid`, and then do the 2 step:
```
let pid, status = Unixext.(waitid P_PID [WEXITED | WNOWAIT] pid) in
Atomic.set need_to_kill false;
let pid, status = Unix.waitpid [] pid in
```